dali/kernels/imgproc/warp/warp_setup.cuh

```diff
@@ -0,0 +1,232 @@
+      span<const TensorShape<ndim>> output_sizes) {
+    TensorListShape<tensor_dim> shape;
+    shape.resize(in_shape.num_samples(), tensor_dim);
+    for (int i = 0; i < in_shape.num_samples(); i++) {
+      int channels = in_shape.tensor_shape_span(i).back();
+      shape.set_tensor_shape(i, shape_cat(output_sizes[i], channels));
+    }
+    return shape;
+  }
+
+
+  template <int shape_ndim>
+  void ValidateOutputShape(
+      const TensorListShape<shape_ndim> &out_shape,
+      const TensorListShape<tensor_dim> &in_shape,
+      span<const TensorShape<ndim>> output_sizes) {
+    TensorListShape<tensor_dim> shape;
+    shape.resize(in_shape.num_samples(), tensor_dim);
+    for (int i = 0; i < in_shape.num_samples(); i++) {
+      int channels = in_shape.tensor_shape_span(i).back();
+      auto out_tensor_shape = out_shape.template tensor_shape<tensor_dim>(i);
+      auto required_tensor_shape = shape_cat(output_sizes[i], channels);
+      DALI_ENFORCE(out_tensor_shape == required_tensor_shape,
+        "Invalid output tensor shape for sample: " + std::to_string(i));
+    }
+  }
+
+  template <typename Backend, typename OutputType, typename InputType>
+  void PrepareSamples(const OutList<Backend, OutputType, tensor_dim> &out,
+                      const InList<Backend, InputType, tensor_dim> &in) {
+    assert(out.num_samples() == in.num_samples());
+    samples_.resize(in.num_samples());
+    for (int i = 0; i < in.num_samples(); i++) {
+      SampleDesc &sample = samples_[i];
+      sample.input = in.tensor_data(i);
+      sample.output = out.tensor_data(i);
+      auto out_shape = out.tensor_shape(i);
+      auto in_shape = in.tensor_shape(i);
+      int channels = out_shape[ndim];
+      sample.channels = channels;
+      sample.out_size = size2vec(out_shape);
+      sample.in_size = size2vec(in_shape);
+
+      sample.out_strides.x = channels;
+      sample.out_strides.y = sample.out_size.x * sample.out_strides.x;
+
+      sample.in_strides.x = channels;
+      sample.in_strides.y = sample.in_size.x * sample.in_strides.x;
+    }
+  }
+
+  dim3 BlockDim() const {
+    return dim3(block_dim_.x, block_dim_.y, block_dim_.z);
+  }
+
+  dim3 GridDim() const {
+    return dim3(grid_dim_.x, grid_dim_.y, grid_dim_.z);
+  }
+
+  span<const SampleDesc> Samples() const { return make_span(samples_); }
+
+  span<const BlockDesc> Blocks() const { return make_span(blocks_); }
+
+
+  ivec<ndim> UniformOutputSize() const {
+    assert(is_uniform_);
+    return uniform_output_size_;
+  }
+
+  ivec<ndim> UniformBlockSize() const {
+    assert(is_uniform_);
+    return uniform_block_size_;
+  }
+
+  bool IsUniformSize() const { return is_uniform_; }
+
+ private:
+  std::vector<SampleDesc> samples_;
+  std::vector<BlockDesc> blocks_;
+  ivec3 block_dim_{32, 32, 1};
+  ivec3 grid_dim_{1, 1, 1};
+  ivec<ndim> uniform_block_size_, uniform_output_size_;
+  bool is_uniform_ = false;
+
+  static inline ivec<ndim> size2vec(const TensorShape<tensor_dim> &size) {
```

Comment:
Maybe `vec` should have some constructor which accepts any data type that has a subscript operator?

Reply:
I think that would create a lot of ambiguity.... Anyway, this function reverses the shape and strips the channel dimension, so it's not exactly a copy.